src/Pool.sol

```diff
@@ -141,157 +157,243 @@ abstract contract Pool is Pausable, Ownable, ERC4626 {
+     * @param amount_ Amount to be withdrawed.
      */
-    function withdraw(uint32 depositId) external {
-        /*
-            1. Checks if the deposit exists.
-            2. Withdraws the deposit (Taking into account the different deposits)
-        */
+    function withdraw(uint128 amount_) external nonReentrant {
+        // Update the rewards per token.
+        updateRewardsPerToken();
+
+        // Check if the user has deposited enough amount as the requested to withdraw.
+        if (_usersData[msg.sender].totalAmountDeposited < amount_) revert Pool_NotEnoughAmountDepositedError();
+
+        // Check user's deposits in order to compute the amount to withdraw.
+        uint128 withdrawAmount_;
+        uint128 totalFee_;
+        for (uint32 i = _usersData[msg.sender].nextDepositIdToRemove; i < _usersData[msg.sender].deposits.length; i++) {
+            // If we haven't reach the total amount requested, we continue.
+            if (withdrawAmount_ >= amount_) {
+                break;
+            }
+
+            // Remaining amount to be obtained from the deposits.
+            uint128 amountRem_ = amount_ - withdrawAmount_;
+            uint128 depositAmount_ = _usersData[msg.sender].deposits[i].amount;
+
+            // Compute the withdraw fee and the resulting amount of the deposit.
+            uint64 withdrawFee_ =
+                uint64(_calcWithdrawFeePerBlock(_usersData[msg.sender].deposits[i].blockNumber) * MULTIPLIER / 100);
+            uint128 depositAmountFeed_ = uint128(depositAmount_ - depositAmount_ * withdrawFee_ / MULTIPLIER);
+
+            // Check if the deposit can manage the full remaining amount of the withdrawal.
+            if (depositAmountFeed_ > amountRem_) {
+                // Update user information regarding the deposit and the total amount deposited.
+                _usersData[msg.sender].deposits[i].amount -= amountRem_;
+                _usersData[msg.sender].totalAmountDeposited -= amountRem_;
+
+                // Update the withdrawAmount and the totalFee.
+                withdrawAmount_ += amountRem_;
+                totalFee_ += uint128((amountRem_ * withdrawFee_) / MULTIPLIER);
+
+                continue;
+            }
+
+            // Update user information regarding the deposit and the total amount deposited.
+            _usersData[msg.sender].deposits[i].amount = 0;
+            _usersData[msg.sender].totalAmountDeposited -= depositAmount_;
+
+            // Update the total deposited amount.
+            _totalAmount -= depositAmount_;
+
+            // Update the withdrawAmount and the totalFee.
+            withdrawAmount_ += depositAmountFeed_;
+            totalFee_ += uint128((depositAmount_ * withdrawFee_) / MULTIPLIER);
+
+            // This deposit it's no longer available. Update the nextDeposit to be removed.
+            _usersData[msg.sender].nextDepositIdToRemove = i == _usersData[msg.sender].deposits.length - 1 ? i : i + 1;
```

Comment:
info: I don't understand why just doing `_usersData[msg.sender].nextDepositIdToRemove = i + 1;` does not work. Let's say, start at deposit 0, delete deposit 0, then next deposit to check is 1 (i + 1).